### ISSUE TITLE
I am curious about an optimization problem

### ISSUE THREAD

**Issue by ASaltyFishBoy** [ReactiveX/RxCpp#523]
I am curious about an optimization problem
When a stream is not used directly, will it be compressed to the next stream?

This problem comes from the following code

observable<MessagePtr> getMessage$()
{
    return m_clientChanged$.get_observable ().scan (observable<Message>()
            , [] (observable<MessagePtr> sendMessage$
                , std::pair<MessagePtr, bool> client)
    {
        if (client.second)
        {
            sendMessage$ = sendMessage$.merge (client.first->getPeerMessage$ ());
        }
        else
        {
            sendMessage$ = sendMessage$.filter ([ = ] (MessageTransferEventPtr message)
            {
                return message->getMessage ()->getTo () == client.first->getServer ();
            });
        }
        return sendMessage$;
    }).switch_on_next();
}

I worry that this code will cause more and more nesting of merge and filter due to the increase of time, causing performance problems

**Comment by kirkshoop** [comment]
Yes, this will be a problem. Use a take_until or take_while before the merge to allow another stream to unsubscribe the client instead of adding a filter to remove the clients messages.
